plonky2/src/hash/hashing.rs

```diff
@@ -101,16 +101,18 @@ pub trait PlonkyPermutation<F: RichField> {
 /// for the hash to be secure, but it can safely be disabled in certain cases, like if the input
 /// length is fixed.
 pub fn hash_n_to_m<F: RichField, P: PlonkyPermutation<F>>(
-    mut inputs: Vec<F>,
+    inputs: &[F],
     num_outputs: usize,
     pad: bool,
 ) -> Vec<F> {
     if pad {
-        inputs.push(F::ZERO);
-        while (inputs.len() + 1) % SPONGE_WIDTH != 0 {
-            inputs.push(F::ONE);
+        let mut padded_inputs = inputs.to_vec();
+        padded_inputs.push(F::ZERO);
+        while (padded_inputs.len() + 1) % SPONGE_WIDTH != 0 {
+            padded_inputs.push(F::ONE);
         }
-        inputs.push(F::ZERO);
+        padded_inputs.push(F::ZERO);
+        return hash_n_to_m::<F, P>(&padded_inputs, num_outputs, false);
```

Comment:
I think we can avoid this cloning too. We process the inputs in one loop as usual, and have a second loop to deal with the padding. But Merkle trees don't use padding, so maybe it doesn't matter.

Reply:
(Discussed offline, decided to hold off for now since we don't use padding in perf sensitive code)